_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+# Spike: Resolve included models
+
+Consider the following domain model(s):
+
+- A model called `Category` with properties `id`, `name`.
+- A model called `Product` with properties `id`, `name`, `categoryId`
+- A `hasMany` relation (Category has many products)
+
+Now consider the following code to retrieve all categories with all related
+products (perhaps to render a home page of a product catalog):
+
+```ts
+categoryRepo.find({include: [{relation: 'products'}]});
+```
+
+## The problem
+
+How to fetch products for each category found?
+
+Additional constraints:
+
+- The target model (`Product`) can be backed by a different database than the
+  source model (`Category`). For example, we can use MongoDB to store categories
+  and MySQL to store products.
+
+- We need to test relations against real databases to ensure we are correctly
+  handling database-specific quirks like:
+  - a limit on the number of items that can be included in `inq`
+  - coercion between `ObjectID` vs. `string` for MongoDB.
+
+## Proposed solution
+
+### 1. Introduce the concept of `InclusionResolver`
+
+An inclusion resolver is a class that can fetch target models for the given list
+of source model instances. The idea is to create a different inclusion resolver
+for each relation type, e.g. `HasManyInclusionResolver`, etc.
+
+```ts
+export interface InclusionResolver {
+  fetchIncludedModels<SourceWithRelations extends Entity>(
+    // list of source models as returned by the first database query
+    sourceEntities: SourceWithRelations[],
+    // inclusion requested by the user (e.g. scope constraints to apply)
+    inclusion: Inclusion,
+    // generic options object, e.g. carrying the Transaction object
+    options?: Options,
+  ): Promise<void>;
+  // ^^ nothing is returned, the source models are updated in-place
+}
```

Comment:
I'm thinking louder here:

1. The generic resolver which can be backed by a property or a function with (optional) args, for example:

`name` -> `obj.name`, `obj.name(...)`, or resolve(obj, 'name', ...) 

2. InclusionResolver extends/implements the generic resolver

Reply:
I'd like to keep InclusionResolver focused on Repository layer and don't deal with specific aspects of different transport protocols (REST, GraphQL, etc.)

To support GraphQL, we can build GraphQL resolvers that use InclusionResolvers as an implementation detail. (I.e. approach this problem in the opposite way from what you suggested.)

You raised a valid point though. I wanted to write an example code showing how a GraphQL resolver can be implemented using `fetchIncludedModels` API and realized that `fetchIncludedModels` is already too coupled with the way how we resolve navigational properties in ORM and REST APIs.

I am going to research this a bit more and try to come up with a simpler & more flexible API.

Reply:
It's also important to ensure GraphQL resolvers are not prone to [SELECT N+1 PROBLEM](https://stackoverflow.com/q/97197/69868). I did a quick search and it looks like the recommended solution is to leverage [DataLoader](https://github.com/graphql/dataloader/) to batch multiple queries into a single one. DataLoader's example showing integration with GraphQL is not trivial: https://github.com/graphql/dataloader#using-with-graphql.

As I see it, implementing inclusion resolvers for GraphQL requires further research that's out of scope of this spike.

Having said that, I still would like to improve the current InclusionResolver API to make it more flexible.

Reply:
Addressed in 2f69523e3.